app/src/lib/shell.ts

```diff
@@ -0,0 +1,107 @@
+/* tslint:disable:no-sync-functions */
+
+import { spawnSync } from 'child_process'
+import * as os from 'os'
+
+const ENVIRONMENT_VARIABLES_TO_PRESERVE = new Set([ 'NODE_ENV', 'NODE_PATH' ])
+
+type IndexLookup = {
+  [propName: string]: string;
+}
+
+/**
+ * Inspect whether the current process is being launched from the macOS tray,
+ * and requires rehydrating of the shell environment variables.
+ *
+ * @param process The current process to inspect.
+ */
+export function shellNeedsPatching(process: NodeJS.Process): boolean {
+  if (__DARWIN__ && !process.env.PWD) {
+    const shell = getUserShell()
+    if (shell.endsWith('csh') || shell.endsWith('tcsh')) {
+      return false
+    }
+    return true
+  }
+
+  return false
+}
+
+/**
+ * Gets a dump of the user's configured shell environment.
+ *
+ * @returns the output of the `env` command or `null` if there was an error.
+ */
+function getRawShellEnv(): string | null {
+  const shell = getUserShell()
+
+  // The `-ilc` set of options was tested to work with the OS X v10.11
+  // default-installed versions of bash, zsh, sh, and ksh. It *does not*
+  // work with csh or tcsh.
+  const results = spawnSync(shell, [ '-ilc', 'env' ], { encoding: 'utf8' })
```

Comment:
```sh
/bin/bash -ilc 'env'  0.40s user 0.26s system 77% cpu 0.853 total
```

It's not great, but wrapping this in a `Promise` and doing it async worried me, especially when we can't `await` at this point of the launch process.